src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/subquery/rewriter/InRewriter.java

```diff
@@ -0,0 +1,117 @@
+/*
+ *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.rewriter.subquery.rewriter;
+
+import com.alibaba.druid.sql.ast.SQLExpr;
+import com.alibaba.druid.sql.ast.expr.SQLBinaryOpExpr;
+import com.alibaba.druid.sql.ast.expr.SQLBinaryOperator;
+import com.alibaba.druid.sql.ast.expr.SQLInSubQueryExpr;
+import com.alibaba.druid.sql.ast.expr.SQLNullExpr;
+import com.alibaba.druid.sql.ast.statement.SQLJoinTableSource.JoinType;
+import com.alibaba.druid.sql.ast.statement.SQLTableSource;
+import com.alibaba.druid.sql.dialect.mysql.ast.statement.MySqlSelectQueryBlock;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.subquery.BlackBoard;
+
+/**
+ * IN Subquery Rewriter.
+ * For example,
+ * SELECT * FROM A WHERE a IN (SELECT b FROM B) and c > 10 should be rewritten to
+ * SELECT A.* FROM A JOIN B ON A.a = B.b WHERE c > 10 and B.b IS NOT NULL.
+ */
+public class InRewriter implements Rewriter {
+    private final SQLInSubQueryExpr inExpr;
+    private final BlackBoard bb;
+    private final MySqlSelectQueryBlock queryBlock;
+
+    public InRewriter(SQLInSubQueryExpr inExpr, BlackBoard bb) {
+        this.inExpr = inExpr;
+        this.bb = bb;
+        this.queryBlock = (MySqlSelectQueryBlock) inExpr.getSubQuery().getQuery();
+    }
+
+    //
+    @Override
+    public boolean canRewrite() {
+        return !inExpr.isNot();
+    }
+
+    /**
+     * Build Where clause from input query.
+     * <p>
+     * With the input query.
+     *         Query
+     *      /    |     \
+     *  SELECT  FROM  WHERE
+     *  |        |    /   |  \
+     *  *         A  c>10 AND INSubquery
+     *                         /    \
+     *                        a    Query
+     *                        /      \
+     *                           SELECT FROM
+     *                             |     |
+     *                             b     B
+     * <p>
+     *
+     */
+    @Override
+    public void rewrite() {
```

Comment:
Want to confirm if I understand right. So either `In` or `Exists` will be created and only 1 `Rewriter` will be invoked. Why we need a global `bb` with stack?

Reply:
It could be multiple In or Exists be created when traverse the AST. we limit the In and Exists to only 1 because other limitation in the implementation, for example, we only support 2 table for join.